### ISSUE TITLE
F to fast forward the highlighted branch

### ISSUE THREAD

**Issue by willparsons** [jesseduffield/lazygit#1895]
**Topic**
It seems like most the commands act on the currently highlighted item except fast-forward which acts on the current checked out branch. 

**Your thoughts**
Imo, this can be a bit annoying and I have accidentally fast-forwarded the current branch expecting it to fast-forward the one I have highlighted.

**Comment by jesseduffield** [comment]
Are you pressing 'f' to do the fast-forwarding? That acts on the selected branch

**Comment by willparsons** [comment]
Yeah I just mean that most actions seem to work on the item you have highlighted except for fast-forwarding. It just feels inconsistent to me.

**Comment by jesseduffield** [comment]
sorry what I meant was that pressing 'f' should work on the highlighted branch. Are you testing this on the latest version?

**Comment by willparsons** [comment]
You're right, it does. Not sure what was happening to me when I was made the issue 😆

**Comment by rcdailey** [comment]
My expectation was a little different -- I am expecting the checked out branch to fast-forward merge the highlighted branch, but it doesn't work that way. This does feel inconsistent to me.

**Comment by stefanhaller** [comment]
@rcdailey Inconsistent with what? I can understand if your expectation was different (although the command's tooltip explains pretty clearly what it does), but what's the inconsistency?

Merging another branch into the checked out branch is done with the `M` command. We don't have an explicit way to do a `--ff` or `--ff-only` merge, but it's coming at some point (see #4091).

**Comment by rcdailey** [comment]
The inconsistency I'm referring to is how `M` and `f` semantically differ:
- When I press `M`, the source branch of a merge is the highlighted branch
- When I press `f`, the source branch of the merge is the tracking branch

However I'd accept the response that my expectations are misplaced. It doesn't seem like the intended design is to have a separate hotkey for every merge type, rather a single key for all merge types (`M`) which *would* make them consistent, and that `f` is just a completely different semantic operation from a UX perspective.

Some of this a new user figuring things out. I typically am not expecting operations to be nested in submenus or modals (requiring two steps). I typically look in the ? menu for direct operations; however not all things are designed this way. So yeah, just learning!

Thanks for the quick response as always!

**Comment by stefanhaller** [comment]
Interesting; I never thought of `f` doing a merge, and if you look in the command log, you'll see that it actually doesn't. I usually keep my command log closed, and when I discovered the `f` command I just took it as some magic shortcut for checking out the branch, pressing `p` on it, and checking out the previous branch again. I didn't really care much about what magic it uses to achieve this, but it never felt like a merge to me.

**Comment by rcdailey** [comment]
Explicit fast forward merges are useful in some workflows. Here's one in particular I do quite often:

1. I create a topic branch off main called `foo`.
2. The purpose of this branch would be to do some quick work to merge directly into `main`.
3. Once I'm done with the work, I rebase it: `git checkout foo && git pull origin`
4. I then switch back to `main`, fast-forward merge `foo` into `main`, then I push: `git checkout main && git merge --ff-only topic && git push`

This may seem a little clunky, but I've developed a habit over the years of avoiding merge commits. My preference for topic branches is to *always rebase* and fast-forward merge back to `main`. This keeps history linear and clean.

So in lazygit, I tried to replicate this flow (starting from step 4 above):

1. press 3 (branches list), hit spacebar on `main`
2. arrow down to `foo`, press `f` (expecting it to fast-forward merge into main, my checked out branch)
3. `P` to push

Step 2 obviously just does a fast-forward of the remote tracking branch but doesn't actually modify my checked out branch. That's where I got confused. I typically configure git to manage remote tracking branches behind the scenes when I `fetch`; I don't ever manage them manually, which is what this `f` functionality appears to support.
